### DISCUSSION TITLE
C++ foxglove websocket server: ambiguous handler invoke timing

### DISCUSSION THREAD

**Discussion by paulsohn** [foxglove/community#1049] [Bugs] [answered]
Hey guys. Thanks for wonderful products and code repositories.
I'm using C++ foxglove websocket server, and found out some features doesn't work out as I expected.

Namely, my server stores the last received message per topic, and if a subscription request from client comes then it emits the saved message of the requested topic immediately. The problem is that `server->sendMessage()` inside of `subscribeHandler()` doesn't work.

I figured out that `sendMessage()` filters out unsubscribed topics, but when `sendMessage()` is called (inside of `subscribeHandler()` the request topic is not yet recognized as a subscribed topic so the call itself has no effect. This is the relevant line:

https://github.com/foxglove/ws-protocol/blob/main/cpp/foxglove-websocket/include/foxglove/websocket/websocket_server.hpp#L1270

In fact, I block-scoped L1271~1272 and move L1270 after the block, then I could get the desired behavior.

Based on its [python](https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/server/__init__.py#L528) and [typescript](https://github.com/foxglove/ws-protocol/blob/main/typescript/ws-protocol/src/FoxgloveServer.ts#L460) counterparts, I suppose that the C++ callback `subscribeHandler()` should be called AFTER the topic has been added to internal subscribed topic table as well. How do you think?

**Note**: Here are the handlers I have checked so far which call timing should be fixed.

* subscribe
* unsubscribe
* client advertise
* client unadvertise
* subscribe connection graph

**Comment by defunctzombie** [comment, marked as answer]
Thanks for the detailed write-up. I've ticketed this internally (https://linear.app/foxglove/issue/FG-9181/c-foxglove-websocket-server-ambiguous-handler-invoke-timing) and we'll try to get to it over the next few weeks. If you feel inspired to make a PR for this before we get around to it we'll happily review.

**Reply by paulsohn** [reply]
Sure, here you go. https://github.com/foxglove/ws-protocol/pull/850
